### PR TITLE
Allocators

### PR DESCRIPTION
**Pull request by SchrodingerZhu**
### What problem does this PR solve?

Problem Summary:
- Add snmalloc
- Update mimalloc 

### What is changed and how it works?
What's Changed:
- Introduced feature flag snmalloc
- Change mimalloc from `miallocator` to `mimalloc`

### Related changes

Internal tickets:
https://internals.tidb.io/t/topic/42/4

### Release note

- Add snmalloc (experimental)
- Update mimalloc crate from `miallocator` to `mimalloc` (experimental)

### CODE REVIEW

File: components/tikv_alloc/Cargo.toml [hunk 1]
```diff
@@ -11,7 +11,7 @@ jemalloc = ["tikv-jemallocator", "tikv-jemalloc-ctl", "tikv-jemalloc-sys"]
 # Build jemalloc's profiling features. Without this
 # certain profile functions will return nothing.
 mem-profiling = ["tikv-jemallocator/profiling", "log"]
-mimalloc = ["mimallocator"]
+snmalloc = ["snmalloc-rs"]
```
Comment (sticnarf) [inline components/tikv_alloc/Cargo.toml:14]: Why the mimalloc feature is directly removed...

Comment (SchrodingerZhu) [reply components/tikv_alloc/Cargo.toml:14]: because this is an optional dependency `mimalloc` which already gives the feature.

Comment (SchrodingerZhu) [reply components/tikv_alloc/Cargo.toml:14]: because this is an optional dependency `mimalloc` which already gives the feature.

Comment (SchrodingerZhu) [reply components/tikv_alloc/Cargo.toml:14]: or if you like I might be able to rename the dependency and rebind `mimalloc` feature flag to it.

Comment (sticnarf) [reply components/tikv_alloc/Cargo.toml:14]: oh never mine. I see the dependency name is changed to `mimalloc` which is just identical to the feature name.

**Comment by SchrodingerZhu** [comment]
I am considering adding benchmarks or tests to this feature. any suggestion？

**Comment by sticnarf** [comment]
> I am considering adding benchmarks or tests to this feature. any suggestion？

@dbsid Could you help test the performance of different allocators?

**Review by sticnarf — APPROVED** [review]
LGTM. Please rebase master

**Comment by sticnarf** [comment]
PTAL @andylokandy

**Review by andylokandy — APPROVED** [review]

**Pull request merged**